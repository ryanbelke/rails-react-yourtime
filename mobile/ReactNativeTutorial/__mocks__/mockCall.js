class MockCall {
  constructor() {
    this.queue = [];
  }

  setMocks(mocks) {
    this.queue = mocks;
  }

  getNextMock() {
    if (!this.queue || !this.queue.length) return undefined;
    return this.queue.shift();
  }

  reset() {
    this.queue = [];
  }
}

const mockCall = new MockCall();

// Mocks calls inside a function under test. This function takes several args and
// stubs the return from call in order of occurence. If no mock were specified
// it returns underfined
export const mockCalls = (...args) => mockCall.setMocks(args);

// Clears all mocks for calls
export const resetMockCalls = () => mockCall.reset();

// The mock call dispatches a fake action to the store with type: 'CALL' and
// function name and args as parameters.
const call = ({ dispatch }) => (f, ...args) => {
  dispatch({ type: 'CALL', name: f.name, args });
  const mock = mockCall.getNextMock();
  const result = typeof mock === 'function' ? mock() : mock;
  return f.then ? Promise.resolve(result) : result;
};

export default call;
