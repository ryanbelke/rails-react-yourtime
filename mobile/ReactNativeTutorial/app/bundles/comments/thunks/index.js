import { Alert } from 'react-native';
import { Actions as navigationActions } from 'react-native-router-flux';
import { reduxUtils } from 'ReactNativeTutorial/app/utils';
import commentsStoreSelector from 'ReactNativeTutorial/app/selectors/commentsStoreSelector';
import commentFormSelector from 'ReactNativeTutorial/app/selectors/commentFormSelector';

import { actions as reduxActions } from 'ReactNativeTutorial/app/reducers';
import * as api from 'ReactNativeTutorial/app/api';

export const fetch = () =>
  async function fetchCommentsEffect(dispatch, _getState, call) {
    dispatch(reduxActions.setLoadingComments(true));
    let response;
    try {
      response = await call(api.fetchComments);
    } catch (e) {
      call(Alert.alert, 'Error', 'Could not connect to server', [{ text: 'OK' }]);
      return;
    } finally {
      dispatch(reduxActions.setLoadingComments(false));
    }
    dispatch(reduxActions.createComments(response.entities.comments));
  };

export const updateForm = reduxActions.updateCommentForm;

export const createComment = () =>
  async function createCommentEffect(dispatch, getState, call) {
    const state = getState();
    const commentsStore = commentsStoreSelector(state);
    const tempId = reduxUtils.getNewId(commentsStore);

    const comment = commentFormSelector(state).merge({ id: tempId }).delete('meta');
    const reduxComment = { [tempId]: comment.toJS() };
    dispatch(reduxActions.createComments(reduxComment));
    call(navigationActions.pop);
    let response;
    try {
      response = await call(api.postComment, comment);
    } catch (e) {
      call(Alert.alert, 'Error', 'Could not post your comment', [{ text: 'OK' }]);
      return;
    } finally {
      dispatch(reduxActions.removeComment(tempId));
    }
    dispatch(reduxActions.createComments(response.entities.comments));
    dispatch(reduxActions.resetCommentForm());
  };
