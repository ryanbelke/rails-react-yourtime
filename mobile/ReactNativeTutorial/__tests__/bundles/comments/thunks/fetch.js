import createDefaultStore from 'redux-mock-store';
import { mockCalls, resetMockCalls } from 'mockCall';
import * as actions from 'ReactNativeTutorial/app/bundles/comments/thunks';

describe('fetch', () => {
  let store;

  beforeEach(() => {
    store = createDefaultStore();
  });

  afterEach(() => {
    resetMockCalls();
  });

  describe('successful fetch', () => {
    it('saves comments into the store', async () => {
      const response = {
        entities: {
          comments: {
            1: { id: 1, author: 'Alexey', text: 'Just some random comment' },
            2: { id: 2, author: 'Justin', text: 'Another random comment' },
          },
        },
      };
      mockCalls(response);
      await store.dispatch(actions.fetch());
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('failing fetch', () => {
    it('dispatches an action with error', async () => {
      mockCalls(() => { throw new Error('Invalid Json'); });
      await store.dispatch(actions.fetch());
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
