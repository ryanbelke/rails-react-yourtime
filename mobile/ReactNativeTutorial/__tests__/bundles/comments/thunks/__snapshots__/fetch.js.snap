exports[`fetch failing fetch dispatches an action with error 1`] = `
Array [
  Object {
    "loading": true,
    "type": "COMMENTS_STORE:SET_LOADING",
  },
  Object {
    "args": Array [],
    "name": "fetchComments",
    "type": "CALL",
  },
  Object {
    "args": Array [
      "Error",
      "Could not connect to server",
      Array [
        Object {
          "text": "OK",
        },
      ],
    ],
    "name": "alert",
    "type": "CALL",
  },
  Object {
    "loading": false,
    "type": "COMMENTS_STORE:SET_LOADING",
  },
]
`;

exports[`fetch successful fetch saves comments into the store 1`] = `
Array [
  Object {
    "loading": true,
    "type": "COMMENTS_STORE:SET_LOADING",
  },
  Object {
    "args": Array [],
    "name": "fetchComments",
    "type": "CALL",
  },
  Object {
    "loading": false,
    "type": "COMMENTS_STORE:SET_LOADING",
  },
  Object {
    "entities": Object {
      "1": Object {
        "author": "Alexey",
        "id": 1,
        "text": "Just some random comment",
      },
      "2": Object {
        "author": "Justin",
        "id": 2,
        "text": "Another random comment",
      },
    },
    "type": "COMMENTS_STORE:CREATE",
  },
]
`;
